proto/protovalidate/buf/validate/validate.proto

```diff
@@ -55,6 +53,81 @@ extend google.protobuf.FieldOptions {
   // Rules specify the validations to be performed on this field. By default,
   // no validation is performed against a field.
   optional FieldConstraints field = 1159;
+
+  // Specifies a shared field constraint rule. This option can be specified on
```

Comment:
I did propose that, and even implemented it in `protovalidate-go` before the current implementation. As a result of many discussions, I wound up backtracking to something very close to @rodaine's original proposal instead. Personally I still like the idea of unifying it all, but I don't think I really convinced anyone else of it.

The main points are:

  - There is concern that using the same extension for shared rules and custom rules would be confusing.

  - It's confusing to understand _when_ shared rules would apply. For example, consider this:

    ```proto
    extend buf.validate.FloatRules {
      bool is_zero [(buf.validate.field).float.const = 0]
    }

    message MyMessage {
      // `is_zero` applies, since it is set.
      float number_1 = 1 [(buf.validate.field).float.(is_zero) = true];
      // `is_zero` still applies, since it is set. Extensions always have field presence.
      float number_2 = 2 [(buf.validate.field).float.(is_zero) = false];
    }
    ```

    This isn't a huge problem for the current version because you can do this:

    ```proto
    extend buf.validate.FloatRules {
      bool is_zero [(buf.validate.shared_field).cel = "rule ? this == 0 : true"]
    }
    ```

  - Moving the standard rules to use the same mechanism as custom constraints will require making changes to how custom constraints work in each runtime to allow rule values to be plumbed through.

If you would prefer a version that eliminates the need for a `SharedFieldConstraints` type and allows using standard constraints recursively, I do have some work already done on that front and can switch gears.

I discussed a design that would work this way here: https://github.com/bufbuild/protovalidate/pull/246#discussion_r1747463918 - but it would need to change at least somewhat, since we can't just blindly `proto.Merge` everything if we want to be able to e.g. bind `rules`/`rule` differently per CEL expression based on where it is in the chain of options.

Reply:
@rodaine I'd love to sync on this for 5 minutes pre-merge

@jchadwick-buf my main comment here is that regardless of design (and even assuming that this existing design stays), the comments/documentation are not clear to me as a user. I'm reading the comments and am not sure how to use protovalidate - this is a critical thing to tackle

Reply:
I understand what you mean, I'll try to improve the comments at least. I do worry that the reason it's hard to explain is because the design is actually a bit confusing, so it would probably be best if we really did make sure that this is what we want to go with.

Reply:
I just noticed a probable huge part of the confusion, which is that I have had a typo here. I need to switch the second one to `SharedFieldConstraints`.

Reply:
Now it is `PredefinedConstraints` and `predefined`. I'm leaving this comment unresolved just in case I've screwed up any details from the meeting.